options.md
# Documentation of all config options

Options are read from the following sources. Options in later sources overwrite previous options.

 1. Defaults from `config.py` module.
 2. `/etc/total-tolles-ferleihsystem.conf`
 3. `./total-tolles-ferleihsystem.conf`
 4. Path specified in `CONFIG_FILE`
 5. Environment Variables


## Main options:

| name                    | ENV                | description | standard value |
|:------------------------|:------------------:|:------------|:--------------:|
| FLASK_APP               | :heavy_check_mark: | Name of Flask APP | `total_tolles_ferleihsystem` |
| MODE                    | :heavy_check_mark: | Mode Selector for `PRODUCTION`, `DEBUG` or `TEST` Environment. Only as env variable! | `PRODUCTION` |
| JWT_SECRET_KEY          | :heavy_check_mark: | Secret key for JWT Tokens. Do NOT make this public! Use a random string! |  |
| SQLALCHEMY_DATABASE_URI | :heavy_check_mark: | Url for Database. [More Info](README.md#install) | `sqlite://:memory:` |
| SQLITE_FOREIGN_KEYS     |                    | Use Sqlite with Foreign Key checks enabled.  | `True` |
| LOG_PATH                | :heavy_check_mark: | Path to log folder. | `/tmp` |
| TMP_DIRECTORY           |                    | | `/tmp` |
| DATA_DIRECTORY          |                    | | `/tmp` |
| CONFIG_FILE             | :heavy_check_mark: | Path to a valid config file (python file). Only as env variable!| `total-tolles-ferleihsystem.conf` |
| CELERY_BROKER_URL       | :heavy_check_mark: | Url for Celery compatible Broker. [More Info](README.md#install) | `amqp://localhost` |
| CELERY_RESULT_BACKEND   | :heavy_check_mark: | Url for Celery compatible result Backend. [More Info](README.md#install) | `rpc://` |
| LOG_FORMAT              |                    | Standard Python log format string. |  |
| AUTH_LOG_FORMAT         |                    | Standard Python log format string. |  |


## Third party options:

 *  [Flask options](http://flask.pocoo.org/docs/1.0/config/#builtin-configuration-values)
 *  [FlaskSQLAlchemy options](http://flask-sqlalchemy.pocoo.org/2.3/config/#configuration-keys)
 *  [Flask JWT-Extended options](https://flask-jwt-extended.readthedocs.io/en/latest/options.html)
 *  [Flask Restplus options](https://flask-restplus.readthedocs.io/en/latest/swagger.html?highlight=RESTPLUS_VALIDATE#the-api-expect-decorator)
 *  [Celery options](http://docs.celeryproject.org/en/latest/userguide/configuration.html)


doc/configuration.md
# System Configuration

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).

The following config sources are all loaded if available. The leading config overrides the latter:

  1. The config file specified in the environment variable `CONFIG_FILE`
  2. The config file `./total-tolles-ferleihsystem.conf`
  3. The config file `/etc/total-tolles-ferleihsystem.conf`
  4. The default values

## Format

A configuration file is a python file which is run when the config is loaded.
This means the format to set a setting is:
```python
key = value
```
The value can be `True` and `False` for boolean values, `0` or `0xff` for integer values and `'/tmp/test'` for string values.

> CAUTION! Be aware that you can potentially run any code inside a config file.
> You shall not allow access to the config from untrusty sources or sources from the internet in order to prevent a remote code execution attack.

For further information on the syntax please read any python tutorial or the [official python specification](https://docs.python.org/3/).

  - [LearnPython.org](https://www.learnpython.org/)
  - [LearnXInYMinutes](https://learnxinyminutes.com/docs/python3/)


## Environment Variables

The api needs two environment variables to function properly, those are `FLASK_APP` and `MODE`.
The `FLASK_APP` variable must be set to `total_tolles_ferleihsystem` as this is required by flask to find and run our api.
The `mode` variable can be set to the following values:

| value        | description |
|:-------------|:------------|
| `PRODUCTION` | This puts the api in production mode, this removes all debug api endpoints and loads different default settings. |
| `DEBUG`      | This puts the api in debug mode, activating more verbose error handling and enables all debug api endpoints. |
| `TEST`       | Currently unused! Will be used for automated testing. |

Additionally the environment variable `FLASK_DEBUG=1` can be provided.
This makes flask reload the api if the source files change, but this is optional.


### Configuration over Environment Variables

Additionally some config options can be set via environment variables.
Of course those settings override every other config file.
This is only possible with the following config options:

  - `JWT_SECRET_KEY` [Details]()
  - `SQLALCHEMY_DATABASE_URI` [Details]()
  - `CELERY_BROKER_URL` [Details]()
  - `CELERY_RESULT_BACKEND` [Details]()



## Configuration Options

### `BCRYPT_HANDLE_LONG_PASSWORDS`

  - Default: `True`
  - Should NOT be changed


### `CELERY_BROKER_URL`

This is the location of the celery broker.

  - Default: `'amqp://localhost'`


### `CELERY_RESULT_BACKEND`

This is the location of the celery result backend.

  - Default: `'rpc://'`


### `DATA_DIRECTORY`

This path must point to a folder.
It is used to store all the files which are uploaded to the system and are available for download.

  - Default: `'/tmp'`


### `DB_UNIQUE_CONSTRAIN_FAIL`

This string is used to match the error string generated by the sql engine from sqlalchemy if a unique constraint has failed.
This is done in order to return the correct error code and error message to the user.
Depending on the engine and driver this needs to be adapted.

  - Default: `UNIQUE constraint failed`


### `JWT_ACCESS_TOKEN_EXPIRES`

This setting sets if the jwt access token and refresh token do expire.

  - Default:
    - Production: `True`
    - Debug: `False`


### `JWT_CLAIMS_IN_REFRESH_TOKEN`

  - Default: `True`
  - Should NOT be changed


### `JWT_SECRET_KEY`

This sets the jwt secret key.
The secret key is used to create the jwt keys thus keeping logins active through restarts or multiple api instances with the same secret key.
If it is changed all previous logins are automatically invalidated.
This should be a secure random string.

  - Default:
    - Production: `''.join(hex(randint(0, 255))[2:] for i in range(16))`
    - Debug: `'debug'`
  - Should always be changed to a non changing value to keep logins active.


### `LOGIN_PROVIDERS`

See [loginproviders.md](loginproviders.md)


### `RESTPLUS_VALIDATE`

  - Default: `True`
  - Should NOT be changed


### `SQLALCHEMY_DATABASE_URI`

This configuration option sets the database which is used to store the data created and used by the system.
Although sqlalchemy might support more options for the database uri then presented below, we only test and develop for the following values:

  - `'sqlite://:memory:'`

    This is the simplest and most unreliable solution.
    This creates a in-memory sqlite database to use, thus it is deleted when the api process terminates.

  - `'sqlite:///<path>'`

    This creates a sqlite database at the specified path.
    This file can be anywhere but be aware of file ownership and and access rights.
    The path should always be absolut.

  - `'mysql+<driver>://<user>:<password>@<host>:<port>/<database>'`

    This allows the api to connect to a mysql or mariadb server.
    The driver typically is either `mysqlconnector` or `cymysql`.
    A complete list cam be found in the [offical documentation](https://docs.sqlalchemy.org/en/13/dialects/mysql.html#dialect-mysql).
    You should make sure that strings are always treated as utf8mb4, otherwise smilies, cjk-characters and other unicode characters may break the api.
    Of course you need to replace `<driver>`, `<user>`, `<password>`, `<host>`, `<port>` and `<database>` with the appropriate values from your database server.

More detailed information about this or other database option can be found on the [official documentation](https://flask-sqlalchemy.palletsprojects.com/en/2.x/config/#connection-uri-format)
Keep in mind that right now you need to setup the database explicitly with `flask db upgrade`.
This makes using an in-memory sqlite db practically impossible to use, since on termination of the db creation command the db is destructed again.

  - Default:
    - Production: `'sqlite://:memory:'`
    - Debug: `'sqlite:////tmp/test.db'`
  - Must be changed in production mode


### `SQLALCHEMY_TRACK_MODIFICATIONS`

If set to True, Flask-SQLAlchemy will track modifications of objects and emit signals.
The library default is None, which enables tracking but issues a warning that it will be disabled by default in the future.
This requires extra memory and should be disabled if not needed.

  - Default: `False`
  - Should NOT be changed


### `TMP_DIRECTORY`

This path must point to a folder.
It is used to store files temporarily during file upload and zip file creation.

  - Default: `'/tmp'`


### `WEBPACK_MANIFEST_PATH`

Pointer to the manifest of webpack

  - Default: `'./build/manifest.json'`


total_tolles_ferleihsystem/config.py
"""Module containing default config values."""

from random import randint
import logging


class Config(object):
    DEBUG = False
    TESTING = False
    RESTPLUS_VALIDATE = True
    BCRYPT_HANDLE_LONG_PASSWORDS = True
    JWT_CLAIMS_IN_REFRESH_TOKEN = True
    JWT_SECRET_KEY = ''.join(hex(randint(0, 255))[2:] for i in range(16))
    CORS_ORIGINS = []
    CORS_SUPPORTS_CREDENTIALS = True
    SQLALCHEMY_DATABASE_URI = 'sqlite://:memory:'
    SQLALCHEMY_TRACK_MODIFICATIONS = False
    DB_UNIQUE_CONSTRAIN_FAIL = 'UNIQUE constraint failed'
    LOGGING = {
        'version': 1,
        'formatters': {
            'extended': {
                'format': '%(asctime)s [%(levelname)s] [%(name)-16s] %(message)s <%(module)s, \
                 %(funcName)s, %(lineno)s; %(pathname)s>',
            },
            'short': {
                'format': '[%(asctime)s] [%(levelname)s] [%(name)-16s] %(message)s',
            }
        },
        'handlers': {
            'default': {
                'class': 'logging.handlers.RotatingFileHandler',
                'formatter': 'extended',
                'filename': '/tmp/ttf-default.log',
                'maxBytes': 104857600,
                'backupCount': 10,
            },
            'auth': {
                'class': 'logging.handlers.RotatingFileHandler',
                'formatter': 'short',
                'filename': '/tmp/ttf-auth.log',
                'maxBytes': 104857600,
                'backupCount': 10,
            },
            'query': {
                'class': 'logging.handlers.RotatingFileHandler',
                'formatter': 'short',
                'filename': '/tmp/ttf-querys.log',
                'maxBytes': 104857600,
                'backupCount': 2,
            },
            'lending': {
                'class': 'logging.handlers.RotatingFileHandler',
                'formatter': 'short',
                'filename': '/tmp/ttf-lending.log',
                'maxBytes': 104857600,
                'backupCount': 100,
            },
            'performance': {
                'class': 'logging.handlers.RotatingFileHandler',
                'formatter': 'short',
                'filename': '/tmp/ttf-performance.log',
                'maxBytes': 104857600,
                'backupCount': 2,
            },
            'console': {
                'class' : 'logging.StreamHandler',
                'formatter': 'extended',
            }
        },
        'loggers': {
            'flask.app.auth': {
                'level': logging.INFO,
                'propagate': False,
                'handlers': ['auth'],
            },
            'flask.app.db': {
                'level': logging.WARNING,
                'propagate': False,
                'handlers': ['query'],
            },
            'sqlalchemy': {
                'level': logging.WARNING,
                'propagate': False,
                'handlers': ['query'],
            },
            'ttf.lending': {
                'level': logging.INFO,
                'propagate': False,
                'handlers': ['lending'],
            },
            'ttf.tasks': {
                'level': logging.WARNING,
                'propagate': False,
                'handlers': ['default'],
            },
            'ttf.performance': {
                'level': logging.WARNING,
                'propagate': False,
                'handlers': ['performance']
            }
        },
        'root': {
            'level': logging.WARNING,
            'handlers': ['default'],
        },
        'disable_existing_loggers': True,
    }
    CELERY_BROKER_URL = 'amqp://localhost',
    CELERY_RESULT_BACKEND = 'rpc://'
    TMP_DIRECTORY = '/tmp'
    DATA_DIRECTORY = '/tmp'

    LOGIN_PROVIDERS = ['Basic']

    BASIC_AUTH_ADMIN_PASS = ''.join(hex(randint(0, 255))[2:] for i in range(8))
    BASIC_AUTH_MOD_PASS = ''.join(hex(randint(0, 255))[2:] for i in range(8))

    LDAP_URI = ""
    LDAP_PORT = 0
    LDAP_SSL = False
    LDAP_START_TLS = False
    LDAP_USER_SEARCH_BASE = ""
    LDAP_GROUP_SEARCH_BASE = ""
    LDAP_USER_RDN = ""
    LDAP_USER_UID_FIELD = ""
    LDAP_GROUP_MEMBERSHIP_FIELD = ""
    LDAP_MODERATOR_FILTER = ""
    LDAP_ADMIN_FILTER = ""
    LDAP_MODERATOR_GROUP_FILTER = ""
    LDAP_ADMIN_GROUP_FILTER = ""

    MONITOR_REQUEST_PERORMANCE = True
    LONG_REQUEST_THRESHHOLD = 1

    JSON_SORT_KEYS = False
    JSONIFY_PRETTYPRINT_REGULAR = False
    RESTPLUS_JSON = {'indent': None}


class ProductionConfig(Config):
    pass


class DebugConfig(Config):
    DEBUG = True
    SQLALCHEMY_DATABASE_URI = 'sqlite:////tmp/test.db'
    LONG_REQUEST_THRESHHOLD = 0
    JWT_SECRET_KEY = 'debug'
    JWT_ACCESS_TOKEN_EXPIRES = False
    CORS_ORIGINS = ["http://localhost:*", "http://127.0.0.1:*"]
    LOGIN_PROVIDERS = ['Debug']
    Config.LOGGING['loggers']['flask.app.auth']['level'] = logging.DEBUG
    Config.LOGGING['loggers']['flask.app.db']['level'] = logging.DEBUG
    Config.LOGGING['loggers']['sqlalchemy.engine'] = {
        'level': logging.WARN,
        'propagate': False,
        'handlers': ['query'],
    }
    Config.LOGGING['loggers']['ttf.lending']['level'] = logging.DEBUG
    Config.LOGGING['loggers']['ttf.tasks']['level'] = logging.DEBUG
    Config.LOGGING['loggers']['ttf.performance']['level'] = logging.DEBUG
    Config.LOGGING['loggers']['flask.app.auth']['level'] = logging.DEBUG

    Config.LOGGING['root']['handlers'].append('console')
    Config.LOGGING['loggers']['ttf.performance']['handlers'].append('console')


class TestingConfig(Config):
    TESTING = True
